File: 8_Heap.cpp
#include <vector>
#include <iostream>
#include <queue>
#include <stack>
#include <unordered_map>
#include <string>
#include <algorithm>
#include <list>
#include <sstream>
#include <climits>
#include <unordered_set>

using namespace std;

/*
1. Kth Largest Element in a Stream: https://leetcode.com/problems/kth-largest-element-in-a-stream/description/
2. Last Stone Weight: https://leetcode.com/problems/last-stone-weight/description/
3. K Closest Points to Origin: https://leetcode.com/problems/k-closest-points-to-origin/description/
4. Kth Largest Element in an Array: https://leetcode.com/problems/kth-largest-element-in-an-array/description/
5. Task Scheduler: https://leetcode.com/problems/task-scheduler/description/
6. Design Twitter: https://leetcode.com/problems/design-twitter/description/
7. Find Median from Data Stream: https://leetcode.com/problems/find-median-from-data-stream/description/
*/

// 1
class KthLargest
{
    /*
        We will use min heap and maintain it's size to be k, so that the Kth largest is at the top of heap
        T:O(logk) S:O(k)
    */
    priority_queue<int, vector<int>, greater<int>> pq_min;
    int size;

    // this will make sure the heap size is never greater than k
    void add_and_maintain_size(const int &val)
    {
        // Each insertion will take O(logk) time
        pq_min.push(val);
        if (pq_min.size() > size)
            pq_min.pop();
    }

public:
    KthLargest(int k, vector<int> &nums)
    {
        size = k;
        for (auto val : nums)
            add_and_maintain_size(val);
    }

    int add(int val)
    {
        add_and_maintain_size(val);
        return pq_min.top();
    }
};

// 2
class LastStone
{
    /*
        We will just simulate the game rules, pick top 2 heaviest and smash them and add it back to remaining stones
        to get top 2 heaviest stones, with multiple insertions while playing thr game we will use max Heap

        T:O(nlogn) S:O(n)
    */
public:
    int lastStoneWeight(vector<int> &stones)
    {
        priority_queue<int> pq_max;
        for (auto s : stones)
            pq_max.push(s);

        while (pq_max.size() > 1)
        {
            auto y = pq_max.top();
            pq_max.pop();
            auto x = pq_max.top();
            pq_max.pop();
            // if the weights are same no need to push to queue as both the stones are destroyed
            if (y != x)
                pq_max.push(y - x); // y will be >= x
        }
        // number of stones remaining can be 0 or 1
        return pq_max.empty() ? 0 : pq_max.top();
    }
};

// 3
class KClosest
{
    /*
        Distance will be ((x-0)^2+(y-0)^2)^0.5, so will just store x^2+y^2 value for distance
        Use max heap of size k so that kth closest will be at the top
        T:O(Nlogk)
        S:O(k)
    */
public:
    vector<vector<int>> kClosest(vector<vector<int>> &points, int k)
    {
        vector<vector<int>> ans;
        priority_queue<pair<int, int>> pq_max;

        // push in max Heap and maintain size k
        for (int i = 0; i < points.size(); i++)
        {
            auto p = points[i];
            // pushing {distance,index}
            pq_max.push({p[0] * p[0] + p[1] * p[1], i});
            if (pq_max.size() > k)
                pq_max.pop();
        }

        // retrieve k closest points using the index
        while (!pq_max.empty())
        {
            ans.push_back(points[pq_max.top().second]);
            pq_max.pop();
        }
        return ans;
    }
};

// 4
class KthLargestInArray
{
    /*
    1. QUICK SELECT METHOD
        T:O(n) Avg case , S:O(1)
        Two Optimizations to avoid O(n^2) worst case scenario:
            1. Use random value for pivot
            2. Use three way parition, this is when multiple same values are present in array

    2. USING MIN HEAP
        T:O(nlogk) S:O(k)
    */
    int generate_random(int s, int e)
    {
        return s + rand() % (e - s + 1);
    }
    // this is same Sort Colors problem: https://leetcode.com/problems/sort-colors/description/
    pair<int, int> threeWayPartition(vector<int> &nums, int s, int e, int &pivot)
    { /*
          2nd Optimization:
          this function will return the start and end position of pivot value, all elements with value=pivot will be present in correct index as if they will be present in sorted array

          till l-1 we have <pivot
          from r+1 we have >pivot
          from [l to r] we have pivot
      */
        int l = s, r = e, i = s;
        while (i <= r)
        {
            if (nums[i] == pivot)
                i++;
            else if (nums[i] < pivot)
            {
                swap(nums[l], nums[i]);
                l++;
                i++;
            }
            else
            {
                swap(nums[r], nums[i]);
                r--;
            }
        }
        return {l, r};
    }

public:
    int findKthLargest_priority_queue(vector<int> &nums, int k)
    {
        priority_queue<int, vector<int>, greater<int>> pq_min;
        for (auto v : nums)
        {
            pq_min.push(v);
            if (pq_min.size() > k)
                pq_min.pop();
        }
        return pq_min.top();
    }
    int findKthLargest(vector<int> &nums, int k)
    {
        srand(time(0));
        int n = nums.size();
        return findPivotValueAtPosition(nums, 0, n - 1, n - k);
    }
    int findPivotValueAtPosition(vector<int> &nums, int s, int e, int pos)
    {
        // 1. Optimization: random pivot
        int pivot = nums[generate_random(s, e)];
        pair<int, int> pivotPos = threeWayPartition(nums, s, e, pivot);

        // found the value which should be present at pos according to sorted array
        if (pivotPos.first <= pos && pos <= pivotPos.second)
            return pivot;
        else if (pos < pivotPos.first) // go to left half
            return findPivotValueAtPosition(nums, s, pivotPos.first - 1, pos);
        else // go to right half
            return findPivotValueAtPosition(nums, pivotPos.second + 1, e, pos);
    }
};

// 5
class Solution
{
    /*
        T:O(nlog26)=>O(n)
        S:O(26) O(1) PQ of max size 26

        In case we want to print tasks in order, we need to create custom comparator with {freq,i}
        such that if we have A:10, B:10 and we pop from top the returned order is A,B
        and when we push back A:9 B:9 the return order again should be A,B
    */
public:
    int leastInterval(vector<char> &tasks, int n)
    {
        // get the frequency of each task
        vector<int> task_freq(26, 0);
        for (auto task : tasks)
            task_freq[task - 'A']++;

        // put the task in max heap
        priority_queue<int> pq_max;
        for (int i = 0; i < 26; i++)
            if (task_freq[i] > 0)
                pq_max.push(task_freq[i]);

        int total_time = 0;
        int idle_time = 0;

        //  everytime we pop, we performed one operation so increase the time
        while (!pq_max.empty())
        {
            auto task_freq = pq_max.top();
            pq_max.pop();
            task_freq--;
            total_time++;

            int no_of_tasks_left = pq_max.size();
            // if we have more tasks in pq than n, then we can execute them and no idle time is needed
            idle_time = max(n - no_of_tasks_left, 0);

            vector<int> task_freqs;
            if (task_freq > 0)
                task_freqs.push_back(task_freq);

            // we will get next n tasks if present, else there will be idle time
            for (int i = 0; i < min(n, no_of_tasks_left); i++)
            {
                auto top_freq = pq_max.top();
                pq_max.pop();
                top_freq--;
                total_time++;
                if (top_freq > 0)
                    task_freqs.push_back(top_freq);
            }

            // push the freq of remaining task back to heap
            for (auto t : task_freqs)
                pq_max.push(t);

            // no tasks remaining, so no additional idle time is needed
            if (pq_max.empty())
                break;

            total_time += idle_time;
        }
        return total_time;
    }
};

// 6
class Twitter
{
    /*
        time variable will be used to know which tweet is latest int time;
        S:O(users+tweets)
        T: next to each funtion
    */
    int time;
    unordered_map<int, unordered_set<int>> follows;
    unordered_map<int, vector<pair<int, int>>> tweets; // user->{time,tweets}
public:
    Twitter()
    {
        this->time = 0;
    }

    void postTweet(int userId, int tweetId)
    {
        // T:O(1)
        tweets[userId].push_back({time++, tweetId});
    }

    vector<int> getNewsFeed(int userId)
    {
        /*
            T: users*10*log10=> O(users)
            we will check 10 recent tweets of people user is following, worst case they can follow all users
        */
        vector<int> feed;
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq_min;

        // own tweets
        for (int i = tweets[userId].size() - 1, count = 0; i >= 0 && count < 10; i--, count++)
            pq_min.push(tweets[userId][i]);

        // tweets from user they are following, start from end as the latest tweets are inserted at end
        for (auto user_following : follows[userId])
            for (int i = tweets[user_following].size() - 1, count = 0; i >= 0 && count < 10; i--, count++)
            {
                // no need more tweet from this user, since heap already has 10 tweets which are newer than current user
                if (pq_min.size() >= 10 && pq_min.top().first >= tweets[user_following][i].first)
                    break;

                pq_min.push(tweets[user_following][i]);
                if (pq_min.size() > 10)
                    pq_min.pop();
            }

        // retrieve the 10 tweets from heap
        while (!pq_min.empty())
        {
            feed.push_back(pq_min.top().second);
            pq_min.pop();
        }
        // the latest should be first, min priority queue had the least recent of the 10 tweets on top
        reverse(feed.begin(), feed.end());
        return feed;
    }

    void follow(int followerId, int followeeId)
    {
        // T:O(1)
        follows[followerId].insert(followeeId);
    }

    void unfollow(int followerId, int followeeId)
    {
        // T:O(1)
        follows[followerId].erase(followeeId);
    }
};

// 7
class MedianFinder
{
    /*
        We will divide the stream into two halfs but we need them in sorted order so we will be using two priority queues,
        1st half will be in pq_max
        2nd half will be in pq_min

        we will maintain pq_max.top() <= pq_min.top()
        S:O(n)
        T: next to each funtion
    */

    priority_queue<int> pq_max;
    priority_queue<int, vector<int>, greater<int>> pq_min;

public:
    MedianFinder() {}

    void addNum(int num)
    {
        // T:O(logn)
        // we will always increase the size of first half i.e pq_max when a number is added
        // to maintain pq_max.top() <= pq_min.top(), we will first insert in 2nd half(pq_min) and from there remove to top element and push in pq_max
        pq_min.push(num);
        pq_max.push(pq_min.top());
        pq_min.pop();

        // if size difference is 2 make both PQs of same size
        if (pq_max.size() - pq_min.size() == 2)
        {
            pq_min.push(pq_max.top());
            pq_max.pop();
        }
    }

    double findMedian()
    {
        // T:O(1)
        // the number of elements are even
        if (pq_max.size() == pq_min.size())
            return (pq_max.top() + pq_min.top()) / 2.0;

        return pq_max.top();
    }
};

File: 8_heap.py

from typing import List 
from collections import defaultdict, OrderedDict, Counter, deque
import heapq

# 1. https://leetcode.com/problems/kth-largest-element-in-a-stream/
class KthLargest:
    
    # Pattern: min heap of size k, as k largest element 
    def __init__(self, k: int, nums: List[int]):
        self.min_heap=[] 
        self.size=k
        for n in nums:
            self.add(n)
    
    # T:O(logk) and S:O(k)
    def add(self, val: int) -> int:    
        heapq.heappush(self.min_heap,val)
        if len(self.min_heap)>self.size:
            heapq.heappop(self.min_heap)
        return self.min_heap[0]

# 2. https://leetcode.com/problems/last-stone-weight/
def lastStoneWeight(stones: List[int]) -> int:
    # T:O(nlogn) and S:O(n)
    
    # Pattern: max heap and how to use python's min heap as max heap
    max_heap=[(-s,s) for s in stones] # store both negative and original value to avoid confusion while popping and pushing back
    heapq.heapify(max_heap) # O(n)
    
    # n*logn 
    while len(max_heap)>1: # we need at least 2 stones to smash
        top1=heapq.heappop(max_heap)[1] #Log(n)
        top2=heapq.heappop(max_heap)[1] #Log(n) 
        weight=top1-top2
        if weight>0:
            heapq.heappush(max_heap,(-weight,weight)) #Log(n)

    return max_heap[0][1] if max_heap else 0

# 3. https://leetcode.com/problems/k-closest-points-to-origin/
def kClosest( points: List[List[int]], k: int) -> List[List[int]]:
    # T:O(nlogk) and S:O(k)
    
    # Pattern: max heap as k smallest distance and how to use python's min heap as max heap
    max_heap=[]
    for i,p in enumerate(points):
        dist=p[0]*p[0]+p[1]*p[1]
        heapq.heappush(max_heap,(-dist,i))
        if len(max_heap)>k:
            heapq.heappop(max_heap)
        
    return [points[i[1]] for i in max_heap] # Use the index stored in heap to get the point from original list

# 4. https://leetcode.com/problems/kth-largest-element-in-an-array/
def findKthLargest(self, nums: List[int], k: int) -> int:
    # T:O(nlogk) and S:O(k)
    # Pattern: min heap of size k, as kth largest element
    min_heap=[]
    for num in nums:
        heapq.heappush(min_heap,num)
        if len(min_heap)>k:
            heapq.heappop(min_heap)
    return min_heap[0]


# 5. https://leetcode.com/problems/task-scheduler/ 
class Solution:
    def leastInterval_new(self, tasks: List[str], n: int) -> int:
        # T:O(n*logk)=>O(n) as k is 26 and S:O(k) where k is the number of unique tasks
        freq = Counter(tasks)
        #freq, ord, key, if B==4, A==3 after reducing B to 3, it should be still B then A in heap for correct ordering
        max_heap = [(-count, i, task) for i, (task, count) in enumerate(freq.items())]
        heapq.heapify(max_heap) # O(k) where k is 26
        
        cycle_size = n + 1
        schedule = []

        while max_heap:

            heap_size = len(max_heap)
            # Review 1: if cycle_size=n+1 < heap size, don't pop all the heap elements, else all type of tasks will be done in this cycle
            heap_tasks_to_do_in_cur_cycle = min(heap_size, cycle_size)  
            reinsert = []

            for _ in range(heap_tasks_to_do_in_cur_cycle):
                count, i, task = heapq.heappop(max_heap)
                schedule.append(task)
                if count + 1 != 0: # Add +1 since freq is in negative
                    reinsert.append((count + 1, i, task))
            
            # Re insert in heap
            for item in reinsert:
                heapq.heappush(max_heap, item)

            # Review 2: Add idle time only if items remaining
            if max_heap:
                idle_slots = cycle_size - heap_size
                for _ in range(idle_slots):
                    schedule.append("")

        return len(schedule)
                
    # we will go cycle by cycle of size n+1, in each cycle we will try to fill the window, if we can't fill the window, we will add idle time
    def leastInterval(self, tasks: List[str], n: int) -> int:
        # T:O(n*logk) and S:O(k) where k is the number of unique tasks
        freq=Counter(tasks)
        max_heap=[-v for v in freq.values()]
        heapq.heapify(max_heap)
        time_taken=0

        while max_heap:
            window=n+1
            rem=[]
            while window and max_heap:
                freq=heapq.heappop(max_heap)
                time_taken+=1
                freq+=1 #freq is in negative so we need to add it
                if freq:
                    rem.append(freq)
                window-=1
            if window!=0 and rem:
                time_taken+=window
                window=0
            for task in rem:
                heapq.heappush(max_heap,task)

        return time_taken

# 6. https://leetcode.com/problems/design-twitter/description/
class Twitter:

    def __init__(self):
        self.user_tweets=defaultdict(list) # userId -> list of (tweet_time,tweetId)
        self.user_follows=defaultdict(set) # userId -> set of followeeId
        self.tweet_time=0

    def postTweet(self, userId: int, tweetId: int) -> None:
        # T:O(1)
        self.user_tweets[userId].append((self.tweet_time,tweetId))
        self.tweet_time+=1
    
    # Pattern: Merge K sorted lists using a heap(special case: 1) decreasing order 2) till specified size)
    def _get_latest_tweets(self,users,size=10):
        # T:O(K+ KlogU) =>O(KlogU) where K is number of tweets to retrieve and U is number of users to merge, S:O(U) for the heap

        max_heap=[]
        latest_tweets=[]

        # Create max_heap of latest tweet from all users | T:O(U) to create the heap and heapify is O(U) as well
        for userId in users:
            if self.user_tweets[userId]:
                tweet_time=self.user_tweets[userId][-1][0]
                idx=len(self.user_tweets[userId])-1
                max_heap.append((-tweet_time,idx,userId))
        heapq.heapify(max_heap)

        # Retreive merged tweets | T:O(KlogU) where K is number of tweets to retrieve and U is number of users to merge
        while len(latest_tweets)!=size and max_heap:
            _,idx,userId=heapq.heappop(max_heap)
            tweet=self.user_tweets[userId][idx][1]
            latest_tweets.append(tweet)
            idx=idx-1
            if idx>=0:
                tweet_time=self.user_tweets[userId][idx][0]
                heapq.heappush(max_heap,(-tweet_time,idx,userId))
                
        return latest_tweets

    def getNewsFeed(self, userId: int) -> List[int]:
        # T:O(klogu) where k is number of tweets to retreive and u is number of users to merge, S:O(u) for the heap
        users=list(self.user_follows[userId])
        users.append(userId)
        return self._get_latest_tweets(users)

    def follow(self, followerId: int, followeeId: int) -> None:
        # T:O(1)
        self.user_follows[followerId].add(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        # T:O(1)
        self.user_follows[followerId].discard(followeeId) # discard is safe to use if followeeId is not in the set, it will not raise error

# 7. https://leetcode.com/problems/find-median-from-data-stream/
class MedianFinder:

    def __init__(self):
        # we need 2 halfs, first half has max on top, second has min on top
        self.max_heap=[] # first half
        self.min_heap=[] # second half
        

    def addNum(self, num: int) -> None:
        # T:O(log(n)) S:O(n)
        # the below 3 lines, increases size of max_heap by 1
        heapq.heappush(self.min_heap,num)
        min_v=heapq.heappop(self.min_heap)
        heapq.heappush(self.max_heap,-min_v)

        # rebalance the heap so that size difference at max is 1
        if len(self.max_heap)-len(self.min_heap)==2:
            max_v=-heapq.heappop(self.max_heap)
            heapq.heappush(self.min_heap,max_v)

    def findMedian(self) -> float:
        # T:O(log(n)) S:O(n)
        if len(self.max_heap)>len(self.min_heap):
            return -self.max_heap[0]
        # return the average, sum of top of both heaps
        v=-self.max_heap[0]
        v+=self.min_heap[0]
        return v/2
